doc/man/man1/thru.1
\"    This documentation is copyright 1994 David Huron.
.TH thru 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +10
.ti -10
\fBthru\fR  --  expand Humdrum abbreviated format representation to through-composed format
.in -10
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +7
.ti -7
\fBthru\fR   [\fB-v  \fIversion\fR]   [\fIinputfile\fR ...]
.in -7
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "thru"
command expands
.I "abbreviated format"
Humdrum representations to
.I "through-composed formats"
in which input passages are rearranged and output according to some specified
.I "expansion list."
.sp 1
.sp 1
Musical scores frequently contain notational devices such as repeat signs
and Da Capos which permit more succinct renderings of a given document.
Humdrum
.I "section labels"
and
.I "expansion-lists"
provide parallel mechanisms for encoding
.I "abbreviated format"
files.
The
.B "thru"
command is normally used to expand various repetition devices.
However, depending on the input, one of several expansions
(dubbed \fIversions\fR) may be possible.
Hence,
.B "thru"
is also useful for selecting a particular edition, performance,
or interpretation from a composite input.
.sp 1
.sp 1
The input to
.B "thru"
must contain one or more sections identified by section labels.
A
.I "section"
is a set of contiguous records.
A
.I "section label"
is a tandem interpretation that consists of a single
asterisk, followed by a greater-than sign, followed by a keyword
that labels the section, e.g.
.sp 1
.sp 1
.in +2
*>Exposition
.br
*>Trio
.br
*>Refrain
.br
*>2nd ending
.br
*>Coda
.in -2
.sp 1
.sp 1
The section label
.I "keywords"
may contain any sequence of the following ASCII characters:
the upper- or lower-case letters \f(CRA-Z\fR, the numbers \f(CR0\fR to \f(CR9\fR,
the underscore (\f(CR_\fR), dash (\f(CR-\fR), period (\f(CR.\fR),
plus sign (\f(CR+\fR), octothorpe (\f(CR#\fR), tilde (\f(CR~\fR),
at-sign (\f(CR@\fR), or space.
All other characters are forbidden.
A section label interpretation may occur anywhere in a Humdrum input,
however, if more than one spine is present in a passage,
identical section labels must appear concurrently in all spines.
.sp 1
.sp 1
Sections begin with a section label and end when either another
section label is encountered, all spines are assigned new
exclusive interpretations, or all spines terminate.
If there is more than one spine present in a passage,
identical section labels must appear concurrently in all spines.
.sp 1
.sp 1
An
.I "expansion-list"
is a tandem interpretation that
takes the form of a single asterisk followed
by a greater-than sign, followed by an optional label, followed by a list of
section-labels enclosed in square brackets and separated by commas.
Examples are given below.
The first and second expansion-lists identify
two section-labels in their lists.
The last three expansion-lists have been labelled `long,' `1955'
and `Czerny_edition' respectively.
.sp 1
.sp 1
.in +2
*>[section1,2nd ending]
.br
*>long[Exposition,Exposition]
.br
*>1955[Aria]
.br
*>Czerny_edition[refrain]
.in -2
.sp 1
.sp 1
The
.B "thru"
command outputs each section in the order specified in the
expansion list.
If more than one expansion list is present in a file,
then the desired
.I "version"
is indicated on the command line via the
.B "-v"
option.
(See EXAMPLES.)
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "thru"
command provides the following options:
.sp 1
.TS
l l.
\fB-h\fR	displays a help screen summarizing the
	  command syntax
\fB-v \fIversion\fR	expand the encoding according to expansion-
	  list label \fIversion\fR
.TE
.sp 1
Options are specified in the command line.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The following examples illustrate the operation of the
.B "thru"
command.
Consider the following simple file:
.in +2
.sp 1
.TS
l l.
**example	**example
*>[A,B,A,C]	*>[A,B,A,C]
*>A	*>A
data-A	data-A
*>B	*>B
data-B	data-B
*>C	*>C
data-C	data-C
*-	*-
.TE
.sp 1
.in -2
This example contains just three data records -- each of
which has been labelled with its own section label.
The file contains a single unlabelled expansion list which
indicates that `A' section should be repeated between the `B' and `C' sections.
The following command:
.sp 1
.sp 1
.in +2
thru inputfile
.in -2
.sp 1
.sp 1
would produce the following \(odthrough-composed\(cd output:
.in +2
.sp 1
.TS
l l.
**example	**example
*thru	*thru
*>A	*>A
data-A	data-A
*>B	*>B
data-B	data-B
*>A	*>A
data-A	data-A
*>C	*>C
data-C	data-C
*-	*-
.TE
.sp 1
.in -2
Notice that the expansion-list record has been eliminated from the
output.
A \f(CR*thru\fR tandem interpretation is added to all output spines
immediately following each instance of an exclusive interpretation in the input.
If \f(CR*thru\fR tandem interpretations are already present in the input,
they are discarded
(thus, running a file through
.B "thru"
twice will not change the file in any way).
Also notice that there are now two sections in the output sharing the
same label (\f(CR*>A\fR).
This duplication of section-labels is not permitted in
abbreviated-format encodings.
.sp 1
.sp 1
Consider the following more complex example.
Imagine a
.I "Da Capo"
work in which a conventional performance procedes as follows:
a first section (`A') is performed twice, followed by
first and second endings -- labelled `B' and `C' respectively.
A subsequent section ensues (`D'), followed by a return to the
first section (`A').
This first section is played just once, immediately followed by
a final coda section (`E').
.sp 1
.sp 1
Imagine also a hypothetical performance of this work in which
Murray Perahia makes three changes:
Perahia repeats the `D' section,
he repeats the `A' section when returning to the Da Capo
-- re-using the first ending before continuing
to the coda following the repetition.
Finally, Perahia has improvised an introduction to the work.
Both the conventional interpretation and the hypothetical
Perahia interpretation can be represented in the same encoded file
as follows.
.in +2
.sp 1
.TS
l.
**example
*>Perahia[X,A,B,A,C,D,D,A,B,A,E]
*>Conventional[A,B,A,C,D,A,E]
*>X
data-X
*>A
data-A
*>B
!! 1st ending
data-B
*>C
!! 2nd ending
data-C
*>D
data-D
*>E
!! Coda
data-E
*-
.TE
.sp 1
.in -2
The hypothetical Perahia version can be recreated by invoking the command:
.sp 1
.sp 1
.in +2
thru -v Perahia inputfile
.in -2
.sp 1
.sp 1
Alternatively, the \(odconventional\(cd interpretation of the Da Capo
structure could be produced by the command:
.sp 1
.sp 1
.in +2
thru -v Conventional inputfile
.in -2
.sp 1
.sp 1
In each case, the
.B "thru"
command will expand the input file according to the designated
label for the expansion-lists.
Note that there is no limit to the number of expansion-lists
that may appear in a Humdrum file.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
DOS 2.0 and up, with the MKS Toolkit.
OS/2 with the MKS Toolkit.
UNIX systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBstrophe\fR (1),
\fByank\fR (1)
.in -2
.sp 1
.sp 1
.SH "WARNINGS"
.in +2
Humdrum ouput is not guaranteed with the
.B "thru"
command.
In order to assure Humdrum output, it is necessary to have the
same number of active spines at each point where sections are joined
together in the expanded output.
In addition, the exclusive interpretations must match where
sections are joined.
.sp 1
.sp 1
Note that if an expansion list appears after some initial data records, then
.B "thru"
causes the initial material to be output before expanding the
document according to the expansion list.
.sp 1
.sp 1
No two expansion lists can be identified using the same version label.
.sp 1
.sp 1
If two sections contain the same section label in the abbreviated
document, then subsequent sections having the same label are ignored
when expanded by
.B "thru."
.sp 1
.sp 1
Inputs that contain different section labels or expansion-lists
in concurrent spines are illegal and will produce an error.
.in -2


toolkit/awk/thru.awk
#####################################################################
##					THRU.AWK
# Programmed by: Tim Racinsky         Date: May, 1994
# Copyright (c) 1994 David Huron
#
# Modifications:
#    Date:   Programmer:         	Description:
# June 7/94	Tim Racinsky	Modified parse_command to work with getopts
# Dec 12/96	David Huron	Fixed bug with wrong spine terminators after comment.
#
# This program prints out a humdrum file according to section labels and
# an expansion-list interpretation.
#
# Main Functions used
#	parse_command()		process_comments()		process_interpretations()
#	process_data()			process_exclusive()		process_tandem()
#	process_spine_path()	process_section_label()	process_expansion_list()
#	process_thru_record()	print_and_reset()
#
#	VARIABLES:
#
#  -null_interp:  Regular expression for null interpretation record.
#  -exclusive_record:  Regular expression for exclusive interpretation record.
#  -totally_exclusive:  Regular expression for exclusive interpretation record
#   that has an exclusive interpretation in every spine.
#  -tandem_interp:  Regular expression for tandem interpretation name.
#  -tandem_record:  Regular expression for tandem interpretation record.
#  -spine_terminate:  Regular expression for spine-terminating record.
#  -section_label_name:  Regular expression for section label name.
#  -expansion_list_name:  Regular expression for expansion-list name.
#  -thru_record:  Regular expression for *thru interpretation.
#  -options:  holds options specified by the user.
#  -version:  holds version to 'thru' if specified by user.
#  -expansion_list_found:  flag to indicate if the expansion-list has
#					  been found yet.
#  -storing:  flag to indicate if currently storing records in a file.
#  -section_count:  increments each time a new section is stored in a file.
#  -output_file:  holds name of output file (for re-direction).
#  -error:  flag to indicate if an error has occurred.
#  -no_of_sections:  holds number of sections specified in expansion-list.
#  -num_of_spines:  holds number of active spines in the printed output.
#  -first_file:  flag to indicate if first file has been processed yet.
#  -last_file:  holds name of file last processed (or currently if FNR != 1)
#
#	ARRAYS:
#
#  -files:  holds the file name for each corresponding section label.
#  -section_progression:  list of sections to be output (in order).
#
BEGIN {
	USAGE="\nUSAGE: thru -h               (Help Screen)\n       thru "\
		 "[-v version] [file ...]"
	TRUE = 1; FALSE = 0
	FS = OFS = "\t"
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	#
	# Set global regular expressions
	#
	null_interp = "^\\*(\t\\*)*$"
	exclusive_record = "^\\*(\\*([^\t])+)?(\t\\*(\\*([^\t])+)?)*$"
	totally_exclusive = "^\\*\\*([^\t])+(\t\\*\\*([^\t])+)*$"
	tandem_interp = "([vx+^-][^\t]+)|([^*vx+\t^-][^\t]*)"
	tandem_record = "^\\*(" tandem_interp ")?(\t\\*(" tandem_interp ")?)*$"
	spine_terminate = "^\\*-(\t\\*-)*$"
	s_label = "[A-Za-z0-9 _+.#@~-]"
	section_label_name = "(^|\t)\\*>" s_label "+($|\t)"
	expansion_list_name="(^|\t)\\*>" s_label "*\\[" s_label "+(," s_label "+)*\\](\t|$)"
	thru_record = "(^|\t)\\*thru(\t|$)"
	#
	# Set other global variables
	#
	error = FALSE
	expansion_list_found = FALSE
	storing = FALSE
	first_file = TRUE
	options = ""
	version = ""
	last_file = ""
	section_count = 1
	no_of_sections = 0
	#
	# Determine the options specified by the user
	#
	parse_command()
	}
{
#
# If this is the first record finish the last file and re-set variables
#
if (FNR == 1)
	{
	if (first_file)
		{
		first_file = FALSE
		last_file = FILENAME
		}
	else print_and_reset()
	}
#
# Process comments, interpretations, and data depending on the first character
#
if ($0 ~ /^!/) process_comments()
else if ($0 ~ /^\*/) process_interpretations()
else process_data()
}
#
# Print out the sections now
#
END	{
#
# If an error has occurred, simply exit
#
if (error)
	exit
else
	{
	#
	# Loop through the sections listed in the expansion-list
	#
	for (i = 1; i <= no_of_sections; i += 1)
		{
		#
		# If the section was not in the input print an error, otherwise, print
		# the section stored in the file
		#
		if (!(section_progression[i] in files))
			{
			print "thru: ERROR: Section " section_progression[i] " not "\
				 "found in input for file " FILENAME > stderr
			error = TRUE
			exit
			}
		else
			{
			while (getline < files[section_progression[i]] > 0)
				{
				print $0
				# Keep track of the number of active spines
				# so that a proper spine-path terminator
				# can be printed.
				if ($0 !~ /^[\*!]/) num_of_spines = NF
				}
			close(files[section_progression[i]])
			}
		}
	#
	# Print out a terminating record
	#
	if (no_of_sections > 0)
		{
		output_line = "*-"
		for (i = 2; i <= num_of_spines; i += 1) output_line = output_line "\t*-"
		print output_line
		}
	#
	# Close and remove any files which may still exist
	#
	for (i in files)
		{
		close(files[i])
		system("rm " files[i])
		}
	}
}

############################################################################
#				FUNCTIONS USED IN THRU.AWK
############################################################################

############################################################################
##				Function parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Otherwise, process any necessary variables
	#
	if (ARGV[2] != "null")
		{
		options = "v"
		version = ARGV[2]
		}
	ARGV[1] = ARGV[2] = ""
	}

##############################################################################
##					Function process_comments
#	This function processes all global and local comments
#
function process_comments()
	{
	if (!expansion_list_found) print $0
	if (storing) print $0 >> output_file
	}

##############################################################################
##					Function process_interpretations
#	This function processes all interpretation records
#
function process_interpretations()
	{
	if ($0 ~ exclusive_record && $0 !~ null_interp) process_exclusive()
	else if ($0 ~ tandem_interp && $0 !~ null_interp) process_tandem()
	else process_spine_path()
	}

##############################################################################
##					Function process_exclusive
#	This function processes all exclusive interpretation records
#
function process_exclusive(    i,thru_line)
	{
	thru_line = ""
	#
	# Create a record showing the *thru interpretation
	#
	for (i = 1; i <= NF; i += 1)
		{
		if ($i ~ /^\*\*/) thru_line = thru_line "*thru\t"
		else thru_line = thru_line "*\t"
		}
	thru_line = substr(thru_line,1,length(thru_line)-1)
	#
	# If the expansion-list has not been found, print the records
	#
	if (!expansion_list_found)
		{
		print $0
		print thru_line
		}
	#
	# If currently storing records and all spines have exclusive
	# interpretations, stop storing.  Otherwise, print the records
	#
	if (storing)
		{
		if ($0 ~ totally_exclusive)
			{
			close(output_file)
			storing = FALSE
			}
		else
			{
			print $0 >> output_file
			print thru_line >> output_file
			}
		}
	}

##############################################################################
##					Function process_tandem
#	This function processes all tandem interpretation records
#
function process_tandem()
	{
	if ($0 ~ section_label_name) process_section_label()
	else if ($0 ~ expansion_list_name) process_expansion_list()
	else if ($0 ~ thru_record) process_thru_record()
	else
		{
		if (!expansion_list_found) print $0
		if (storing) print $0 >> output_file
		}
	}

#############################################################################
##					Function process_section_label
#	This function processes records that contain section labels
#
function process_section_label(   good_record,label)
	{
	if (!expansion_list_found) print $0
	good_record = TRUE
	#
	# Check that all spines have the same section label
	#
	for (i = 2; i <= NF; i += 1)
		{
		if ($i != $1) { good_record = FALSE; break }
		}
	#
	# If all spines have the same label
	#
	if (good_record)
		{
		#
		# Stop storing if currently storing a section
		#
		if (storing)
			{
			storing = FALSE
			close(output_file)
			}
		label = substr($1,3)
		#
		# If this label has not yet been seen yet, begin storing it in a file
		#
		if (!(label in files))
			{
			"echo $TMPDIR/thru" section_count | getline output_file
			close("echo $TMPDIR/thru" section_count)
			files[label] = output_file
			section_count += 1
			storing = TRUE
			print $0 > output_file
			}
		}
	#
	# Print an error if it is an invalid section label record
	#
	else
		{
		print "thru: ERROR: Invalid section label record at line " \
			 FNR " of file " FILENAME > stderr
		error = TRUE
		exit
		}
	}

#############################################################################
##					Function process_expansion_list
#	This function processes records that contain expansion-lists
#
function process_expansion_list(   label,good_record,list)
	{
	good_record = TRUE
	label = ""
	#
	# Check that all spines have the same expansion-list interpretation
	#
	for (i = 2; i <= NF; i += 1)
		{
		if ($i != $1) { good_record = FALSE; break }
		}
	#
	# If all spines have the same expansion-list interpretation
	#
	if (good_record)
		{
		#
		# Check if there is a 'version' with the expansion-list
		#
		if (match($1,"^\\*>" s_label "+\\[") > 0)
			label = substr($1,RSTART+2,RLENGTH-3)
		#
		# If this is the correct version or new version was specified
		#
		if ((options ~ /v/ && label == version) \
								|| (options !~ /v/ && label == ""))
			{
			#
			# If this is not the first expansion-list for this version, error
			#
			if (expansion_list_found)
				{
				print "thru: ERROR: Repeated expansion-list at line " \
					 FNR " of file " FILENAME > stderr
				error = TRUE
				exit
				}
			#
			# Otherwise, store the sections to be output in the array
			# section_progression
			#
			else
				{
				expansion_list_found = TRUE
				match($1,"\\[" s_label "+(," s_label "+)*\\]")
				list = substr($1,RSTART+1,RLENGTH-2)
				no_of_sections = split(list,section_progression,",")
				}
			}
		}
	#
	# If not a valid record, print an error
	#
	else
		{
		print "thru: ERROR: Invalid expansion-list record at line " \
			 FNR " of file " FILENAME > stderr
		error = TRUE
		exit
		}
	}

############################################################################
##					Function process_thru_record
#	This record processes those records that already contain a thru
# tandem interpretation.
#
function process_thru_record(    i)
	{
	#
	# Delete any instances of *thru found in the input
	#
	for (i = 1; i <= NF; i += 1)
		{
		if ($i == "*thru") $i = "*"
		}
	if ($0 !~ null_interp)
		{
		if (!expansion_list_found) print $0
		if (storing) print $0 >> output_file
		}
	}

##############################################################################
##					Function process_spine_path
#	This function processes all spine_path interpretation records (and null
# interpretation records).
#
function process_spine_path()
	{
	if (!expansion_list_found) print $0
	#
	# If currently storing records and this is a spine-terminating record
	# (i.e. all spines terminate), then stop storing
	#
	if (storing)
		{
		if ($0 ~ spine_terminate)
			{
			storing = FALSE
			close(output_file)
			}
		else print $0 >> output_file
		}
	}

#############################################################################
##					Function process_data
#	This function processes data records
#
function process_data()
	{
	if (!expansion_list_found) print $0
	if (storing) print $0 >> output_file
	}

############################################################################
##				Function print_and_reset
#	This function prints the sections from the previous file and resets
# the variables for the new file.  It is similar to the END section of this
# program except the arrays need to be reset as well.
#
function print_and_reset(    i,temp_line)
	{
	temp_line = $0
	for (i = 1; i <= no_of_sections; i += 1)
		{
		if (!(section_progression[i] in files))
			{
			print "thru: ERROR: Section " section_progression[i] " not "\
				 "found in input for file " last_file > stderr
			error = TRUE
			exit
			}
		else
			{
			while (getline < files[section_progression[i]] > 0) print $0
			close(files[section_progression[i]])
			}
		}
	if (no_of_sections > 0)
		{
		output_line = "*-"
		for (i = 2; i <= NF; i += 1) output_line = output_line "\t*-"
		print output_line
		}
	for (i in files)
		{
		close(files[i])
		system("rm " files[i])
		}
	expansion_list_found = FALSE
	storing = FALSE
	last_file = FILENAME
	section_count = 1
	no_of_sections = 0
	for (i in files) delete files[i]
	for (i in section_progression) delete section_progression[i]
	$0 = temp_line
	}
